tracks/bash/exercises/hamming/mentoring.md

```diff
@@ -0,0 +1,147 @@
+### Reasonable Solutions
+
+```bash
+#!/usr/bin/env bash
+main () {
+  (( "$#" == 2 )) || {
+      echo "Usage: hamming.sh <string1> <string2>"
+      exit 1
+  }
+  
+  # Regular vars are easier to read when doing fancy parameter expansion.
+  a=$1 b=$2 
+  
+  (( "${#a}" == "${#b}" )) || {
+      echo "left and right strands must be of equal length"
+      exit 1
+  }
+
+  declare -i count
+  for (( i = 0; i < "${#a}"; i++ )); do
+    [[ "${a:i:1}" == "${b:i:1}" ]] || count+=1
+  done
+
+  printf '%d' "$count"
+}
+
+main "$@"
+```
+```bash
+#!/usr/bin/env bash
+main () {
+  (( $# != 2 )) && 
+    echo "Usage: hamming.sh <string1> <string2>" && 
+    exit 1
+  (( ${#1} != ${#2} )) && 
+    echo "left and right strands must be of equal length" && 
+    exit 1
+
+  hamming=0;
+  for (( i = 0; i < ${#1}; i++ )); do
+    [[ "${1:i:1}" != "${2:i:1}" ]] && ((++hamming))
+  done
+
+  echo "$hamming"
+}
+main "$@"
+```
+```bash
+#!/usr/bin/env bash
+main () {
+    if (( $# != 2 )); then 
+        echo "Usage: hamming.sh <string1> <string2>";
+        exit 1;
+    fi
+    if (( ${#1} != ${#2} ))
+    then
+        echo "left and right strands must be of equal length";
+        exit 1;
+    fi
+    hamming=0
+    for (( i=0; i<${#1}; i++ )); do
+        [[ "${1:$i:1}" != "${2:$i:1}" ]] && let "hamming+=1"
```

Comment:
Avoid `let`. Prefer `((` or `$((`

Reply:
Disclaimer, I am not an expert - just trying to get some clarification.

Avoid `let`. Prefer `((` or `$((`

Convention is to drop semi-colons - above all keep consistency whether you have semicolons or not.

These recommendations I could only find in one style guide:
https://github.com/bahamas10/bash-style-guide

Other style guides found:
https://google.github.io/styleguide/shellguide.html
https://github.com/azet/community_bash_style_guide

I still have homework to research some of the resources you mentioned:
[Wooledge](https://mywiki.wooledge.org/) or [bash-hackers](https://wiki.bash-hackers.org/)

Is there a style guide that you recommend?

Reply:
I've never heard of bahamas before. Wooledge and Bash Hackers provide solid advice but don't set out to be style guides. There's the Google Style Guide that is mostly pretty solid. If you _need_ a style guide, I'd point to that. If you need solid bash info/advice/recommendations, Wooledge or Bash Hackers are the best places to go.

http://wooledge.org/~greybot/meta/let

Reply:
https://wiki.bash-hackers.org/commands/builtin/let 👍